### PR TITLE
Add stats and firmware Homie extensions, and tests

### PR DESCRIPTION
**Pull request by qwandor-google**

### CODE REVIEW

File: publish-mqtt/src/homie.rs [hunk 1]
```diff
@@ -316,16 +406,22 @@ async fn publish_retained(
...
     use super::*;
+    use async_channel::Receiver;
 
-    #[tokio::test]
-    #[should_panic(expected = "Tried to add node with duplicate ID")]
-    async fn test_duplicate_node() {
-        let (tx, rx) = async_channel::unbounded();
-        let mut device = HomieDevice::new(
-            tx,
+    fn test_device() -> (HomieDevice, Receiver<Request>) {
```
Comment (alsuren) [inline publish-mqtt/src/homie.rs]: We seem to have gone from the python style of naming (where tests are named `test_*()`) to some other scheme, where fixture factories are called `test_*()`. Is this on purpose?

Maybe the factory could be called `make_test_device()`?

When it comes to naming tests, I quite like the jest style of naming (`describe("the function name", ... it("explodes when you try to add a duplicate node", ...) ...)`). Rust can't be that expressive, but you can get quite close. I quite like the descriptiveness of the suggestions in https://dzone.com/articles/7-popular-unit-test-naming but I disagree with the Super_UglyCased_NamingScheme.

If we go for something like 2, we would end up with tests renamed like this:

```
add_node => add_node_succeeds_before_and_after_calling_start
ready_before_start => ready_fails_if_called_before_start
read_removed_node => add_node_succeeds_if_node_has_been_removed
duplicate_node => add_node_fails_if_given_duplicate_id
set_firmware => set_firmware_succeeds_before_start
start_empty => start_succeeds_with_no_sensors
set_firmware_after_start => set_firmware_fails_after_start
```
(ordering is from the output of `cargo test`. I'm not sure what causes it to run things in that order)

File: publish-mqtt/src/homie.rs [hunk 2]
```diff
@@ -171,10 +183,30 @@ impl HomieDevice {
             HOMIE_VERSION,
         )
         .await?;
+        // TODO: Send $localip and $mac too.
```
Comment (alsuren) [inline publish-mqtt/src/homie.rs:186]: is this a TODO for part of this PR, or a later PR? Does it also want to be tracked on the board, or is it fine to stay as just a TODO in the code?

File: publish-mqtt/src/homie.rs [hunk 3]
```diff
@@ -303,6 +344,55 @@ impl HomieDevice {
...
+            format!("{}/$stats/interval", device_base),
+            STATS_INTERVAL.as_secs().to_string(),
+        )
+        .await
+    }
+
+    /// Send latest stats, if enough time has passed since last time.
+    async fn refresh(&mut self) -> Result<(), SendError<Request>> {
+        let now = Instant::now();
+        if now > self.last_sent + STATS_INTERVAL {
```
Comment (alsuren) [inline publish-mqtt/src/homie.rs]: Ideally, I would ask for the initial last_sent to be set to the minimum value that Instant can take, but Rust doesn't seem to expose this. Failing that, let's at least make sure that things work if time stands still (it's common to freeze time in tests in python. I don't know how common this is in rust).
```suggestion
        if now >= self.last_sent + STATS_INTERVAL {
```

File: publish-mqtt/src/homie.rs (hunk 2 above)
Comment (qwandor-google) [reply publish-mqtt/src/homie.rs:186]: For a later PR. Up to you how you want to manage tasks, I can add it if you want.

File: publish-mqtt/src/homie.rs (hunk 1 above)
Comment (qwandor-google) [reply publish-mqtt/src/homie.rs]: Naming tests `test_` seemed a bit unnecessary when they are already in a `tests` module. Renamed, mostly as you suggested.

**Pull request merged**